docs/codegen-tour.md

```diff
@@ -0,0 +1,843 @@
+    /// Encode this type as raw big-endian bytes
+    fn to_raw(self) -> Self::Raw;
+}
+
+/// A wrapper around raw big-endian bytes for some type.
+#[derive(Clone, Copy, PartialEq, Eq)]
+#[repr(transparent)]
+pub struct BigEndian<T: Scalar>(T::Raw);
+```
+
+The `Scalar` trait handles conversion of a type to and from its raw representation
+(a fixed-size byte array) and the `BigEndian` type is way of representing some
+fixed number of bytes, and associating them with a concrete type; it has `get`
+and `set` methods which read or write the underlying bytes, relying on the
+`from_raw` and `to_raw` methods on Scalar.
+
+This is a compromise. The `Raw` associated type is expected to always be a
+fixed-size byte array; say `[u8; 2]` for a `u16`, or `[u8; 3]` for an `Offset24`.
+
+Ideally, the scalar trait would look like,
+
+```rust
+trait Scalar {
+    const RAW_SIZE: usize;
+    fn from_raw(bytes: [u8; Self::RAW_SIZE]) -> Self;
+    fn to_raw(self) -> [u8; Self::RAW_SIZE];
+}
+```
+
+But this is not currently something we can express with Rust's generics.
+
+In any case: what this lets us do is avoid having two separate sets of types for
+the 'raw' and 'native' cases; we have a single wrapper type that we use anytime
+we want to indicate that a type is in its raw form. This has the additonal
+advantage that we can define new types in our generated code that implement
+`Scalar`, and then those types can automatically work with `BigEndian`; this is
+useful for things like custom enums and flags that are defined at various points
+in the spec.
+
+### `FixedSize`
+
+In addition to these two traits, we also have a [`FixedSize`][] trait, which is
+implemented for all scalar types (and later, for structs consisting only of
+scalar types). This trait consists of a single associated constant:
+
+```rust
+/// A trait for types that have a known, constant size.
+pub trait FixedSize: Sized {
+    /// The raw (encoded) size of this type, in bytes.
+    const RAW_BYTE_LEN: usize;
+}
+```
+
+This is implemented for both all the scalar values, as well as all their
+`BigEndian` equivalents; and in both cases, the value of `RAW_BYTE_LEN` is the
+size of the raw (big-endian) representation.
+
+### <a id="read-tables"></a> tables
+
+Conceptually, a table object is additional type information laid over a
+`FontData` object (a wrapper around a rust byte slice (`&[u8]`), essentially
+a pointer plus a length). It provides typed access to that tables fields.
+
+Conceptually, this looks like:
+
+```rust
+pub struct MyTable<'a>(FontData<'a>);
+
+impl MyTable<'_> {
+    /// Read the table's first field
+    pub fn format(&self) -> u16 {
+        self.0.read_at(0)
+    }
+}
+```
+
+In practice, what we generate is slightly different: instead of
+generating a struct for the table itself (and wrapping the data directly)
+we generate a 'marker' struct, which defines the type of the table, and then we
+combine it with the data via a `TableRef` struct.
```

Comment:
I feel like there is a missing "We do this <because reasons>"

Reply:
Hmm, where exactly? I'm staring at this and I'm not seeing it. :/